File: LinkedList.cs
using System;
using System.Collections.Generic;

namespace Stack
{
    class LinkedList<T>
    {
        public Node<T> HeadNode
        { get; private set; }
        public LinkedList(T value = default)
        {
            HeadNode = new Node<T>(value);
        }

        public void InsertBeginning(T value)
        {
            Node<T> newNode = new Node<T>(value);
            newNode.NextNode = HeadNode;
            HeadNode = newNode;
        }

        public string Stringify()
        {
            string s = "";
            Node<T> current = HeadNode;
            while (current != null)
            {
                if (current.Value != null)
                {
                    s += Convert.ToString(current.Value ) + "; ";
                }
                current = current.NextNode;
            }
            return s;
        }

        public static void SwapNodes(LinkedList<T> inputList, T val1, T val2)
        {
            Console.WriteLine($"Swapping {val1} with {val2}");

            Node<T> node1_prev = null;
            Node<T> node2_prev = null;
            Node<T> node1 = inputList.HeadNode;
            Node<T> node2 = inputList.HeadNode;

            if (EqualityComparer<T>.Default.Equals(val1, val2))
            {
                Console.WriteLine("Elements are the same. No swap necessary.");
                return;
            }

            while (node1 is not null)
            {
                if (EqualityComparer<T>.Default.Equals(node1.Value, val1))
                {
                    break;
                }
                node1_prev = node1;
                node1 = node1.NextNode;
            }
            while (node2 is not null)
            {
                if (EqualityComparer<T>.Default.Equals(node2.Value, val2))
                {
                    break;
                }
                node2_prev = node2;
                node2 = node2.NextNode;
            }

            if (node1 is null || node2 is null)
            {
                Console.WriteLine("Swap not possible - one or more elements not in list.");
                return;
            }

            if (node1_prev is null)
            {
                inputList.HeadNode = node2;
            }
            else
            {
                node1_prev.NextNode = node2;
            }

            if (node2_prev is null)
            {
                inputList.HeadNode = node1;
            }
            else
            {
                node2_prev.NextNode = node1;
            }

            Node<T> temp = node1.NextNode;
            node1.NextNode = node2.NextNode;
            node2.NextNode = temp;
        }
    }
}

File: linked_list.rb
require './node.rb'

class LinkedList
    attr_accessor :head_node
    def initialize(value = nil)
        @head_node = Node.new(value)
    end

    def insert_beginning(value)
        new_node = Node.new(value)
        new_node.next_node = @head_node
        @head_node = new_node
    end

    def stringify_list
        string = ""
        current = @head_node
        while !current.nil?
            if current.value != nil
                string += (current.value).to_s + "; "
            end
            current = current.next_node
        end
        return string
    end

    def self.swap_nodes(input_list, val1, val2)
        puts "Swapping #{val1} with #{val2}"

        node1_prev = nil
        node2_prev = nil
        node1 = input_list.head_node
        node2 = input_list.head_node

        if val1 == val2
            puts "Elements are the same. No swap required"
            return
        end

        while !node1.nil?
            if node1.value == val1
                break
            end
            node1_prev = node1
            node1 = node1.next_node
        end

        while !node2.nil?
            if node2.value == val2
                break
            end
            node2_prev = node2
            node2 = node2.next_node
        end

        if node1.nil? or node2.nil?
            puts "Swap not possible - one or more elements not in list"
            return
        end

        if node1_prev.nil?
            input_list.head_node = node2
        else
            node1_prev.next_node = node2
        end

        if node2_prev.nil?
            input_list.head_node = node1
        else
            node2_prev.next_node = node1
        end

        temp = node1.next_node
        node1.next_node = node2.next_node
        node2.next_node = temp
    end
end


